.gitlab-ci.yml

```diff
@@ -125,82 +137,35 @@ build:merge:
     # Runs on tag pipeline where the tag is a prerelease or release version
     - if: $CI_COMMIT_TAG =~ /^v[0-9]+\.[0-9]+\.[0-9]+(?:-.*[0-9]+)?$/
 
-build:linux:
+build:build-generate:
   stage: build
-  needs: []
   script:
+    - mkdir -p ./tmp
     - >
         nix-shell --run '
-        npm run build --verbose;
-        npm test -- --ci --coverage;
+        ./scripts/build-pipelines.sh > ./tmp/build-pipelines.yml
         '
   artifacts:
     when: always
-    reports:
-      junit:
-        - ./tmp/junit/junit.xml
-      coverage_report:
-        coverage_format: cobertura
-        path: ./tmp/coverage/cobertura-coverage.xml
     paths:
-      # Only the build:linux preserves the dist
-      - ./dist
-  coverage: '/All files[^|]*\|[^|]*\s+([\d\.]+)/'
-  rules:
-    # Runs on staging commits and ignores version commits
-    - if: $CI_COMMIT_BRANCH == 'staging' && $CI_COMMIT_TITLE !~ /^[0-9]+\.[0-9]+\.[0-9]+(?:-.*[0-9]+)?$/
-    # Runs on tag pipeline where the tag is a prerelease or release version
-    - if: $CI_COMMIT_TAG =~ /^v[0-9]+\.[0-9]+\.[0-9]+(?:-.*[0-9]+)?$/
-
-build:windows:
-  stage: build
-  needs: []
-  tags:
-    - windows
-  before_script:
-    - choco install nodejs --version=16.14.2 -y
-    - refreshenv
-  script:
-    - npm config set msvs_version 2019
-    - npm install --ignore-scripts
-    - $env:Path = "$(npm bin);" + $env:Path
-    - npm run build --verbose
-    - npm test -- --ci
-  artifacts:
-    when: always
-    reports:
-      junit:
-        - ./tmp/junit/junit.xml
+      - ./tmp/build-pipelines.yml
   rules:
     # Runs on staging commits and ignores version commits
     - if: $CI_COMMIT_BRANCH == 'staging' && $CI_COMMIT_TITLE !~ /^[0-9]+\.[0-9]+\.[0-9]+(?:-.*[0-9]+)?$/
     # Runs on tag pipeline where the tag is a prerelease or release version
     - if: $CI_COMMIT_TAG =~ /^v[0-9]+\.[0-9]+\.[0-9]+(?:-.*[0-9]+)?$/
 
-build:macos:
+build:platforms:
```

Comment:
Since `check:test` became `check:test-generate` & `check:test`, then I can see that the `build:*` jobs have to do have something similar.

It means all the build platforms `build:linux`, `build:macos`.. .etc, all need to be generated, since the shard count will be used.

I'm just wondering if it is better to do:

```
build:linux-generate
build:linux

build:windows-generate
build:windows

build:macos-generate
build:macos
```

So that the generation of each pipeline is specific to the platform.

If you are going to centralise it to `build:platforms`, then I suggest preserving the naming convention:

```
build:platforms-generate
build:platforms
```

Finally once you have done all of this. The build artifacts themselves have to be extracted to the next stage.

Are all of the artifacts extracted to the integration stage?

Reply:
The only artifacts created by those jobs are junit and coverage reports, which are being displayed in gitlab.

Reply:
The `build:linux` job was originally producing the `dist` which is needed for subsequent integration tests. Without that `dist`, the `prerelease` and `release` jobs should not be working properly.

You want to make sure that the `dist` is actually being exported to the subsequent jobs that depend on the parent/child pipeline.

All `build:*` jobs produce the `dist`, but only `build:linux` uploads the `dist`. This was because it wasn't necessary for each job to upload their own dist, and it would be a waste of a job to just build the dist on their own. However it may be clearer to havea job that just builds the `dist` so it can be sent to where it is needed once.

Reply:
Actually the `dist` might only be used by the release and prerelease jobs. The integration jobs are only using the builds coming out of the `release.nix`.

Reply:
So yea, please confirm the existence of `dist` that gets extracted out of the child pipeline, and that goes into the `build:prerelease` or `release:distribution`.

If it doesn't work, we would need to do that building of the distribution elsewhere beforehand. Maybe outside the child pipeline, and then refer to that dist artifact for whatever jobs need it.

Reply:
The `release:distribution` job doesn't seem to have any issues with downloading the artifacts it needs and it succeeds. I also downloaded the artifacts created by the `build:linux` job and it looks like a correct dist. The only issue is that the dist gets created for every shard, so if the `build:linux` job is split into 5 shards then 5 dists will be built and saved. So in this case it might be better to put the creation of the dist into a separate job.

Reply:
Yea, it's not so much that each shard is rebuilding that's a problem, it's that each shard shouldn't be uploading it.

Ok let's use a different job to do the build and pass it into the release/prerelease jobs.

However in ts demo lib native, each build job/test job needs to get their own `npm run build`, because that also builds the native addon which is platform-specific.

So separating out the `npm run build` in this case is just to get the `dist` out to be used by the pre/release jobs.

Reply:
This job can be called `build:dist`.